src/sentry/services/hybrid_cloud/import_export/impl.py

```diff
@@ -0,0 +1,166 @@
+# Please do not use
+#     from __future__ import annotations
+# in modules such as this one where hybrid cloud data models or service classes are
+# defined, because we want to reflect on type annotations and avoid forward references.
+
+from typing import List, Optional
+
+from django.core.exceptions import ValidationError as DjangoValidationError
+from django.core.serializers import deserialize
+from django.core.serializers.base import DeserializationError
+from django.db import DatabaseError, IntegrityError, connections, router, transaction
+from rest_framework.serializers import ValidationError as DjangoRestFrameworkValidationError
+
+from sentry.backup.dependencies import NormalizedModelName, PrimaryKeyMap, get_model, get_model_name
+from sentry.backup.errors import ImportErrorKind
+from sentry.backup.findings import InstanceID
+from sentry.backup.helpers import EXCLUDED_APPS, Filter, ImportFlags
+from sentry.backup.scopes import ImportScope
+from sentry.services.hybrid_cloud.import_export.model import (
+    RpcFilter,
+    RpcImportError,
+    RpcImportResult,
+)
+from sentry.services.hybrid_cloud.import_export.service import ImportExportService
+
+
+class DatabaseBackedImportExportService(ImportExportService):
+    def import_by_model(
+        self,
+        *,
+        model_name: str,
+        scope: Optional[ImportScope] = None,
+        flags: ImportFlags,
+        filter_by: List[RpcFilter],
+        in_pk_map: PrimaryKeyMap,
+        json_data: str,
+    ) -> RpcImportResult:
+        batch_model_name = NormalizedModelName(model_name)
+        model = get_model(batch_model_name)
+        if model is None:
+            return RpcImportError(
+                kind=ImportErrorKind.UnknownModel,
+                on=InstanceID(model=batch_model_name),
+                reason=f"The model `{batch_model_name}` could not be found",
+            )
+        if scope is None:
+            return RpcImportError(
+                kind=ImportErrorKind.UnspecifiedScope,
+                on=InstanceID(model=batch_model_name),
+                reason="The RPC was called incorrectly, please set an `ImportScope` parameter",
+            )
+
+        filters: List[Filter] = []
+        for fb in filter_by:
+            if NormalizedModelName(fb.model_name) == batch_model_name:
+                filters.append(Filter(model, fb.field, fb.values))
+
+        try:
+            using = router.db_for_write(model)
+            with transaction.atomic(using=using):
```

Comment:
@azaslavsky `ImportExportService` can only access tables that reside exclusively in the control silo (because of `local_mode = SiloMode.CONTROL`).

Do you anticipate that `import_export_service` have access to both the region and control silo resource?

If so, we'll want to define separate services for both the region and control silo.

An example can be seen here for `access_service`: https://github.com/getsentry/sentry/blob/c967fcead95f889efa5faff9c512c90613d0d66a/src/sentry/services/hybrid_cloud/access/service.py#L132-L138

Reply:
Importing/exporting will always be run in the regional codepath when in siloed mode. So the only time we need to "go over the wire" will be when we are trying to import/export control models. Is there anything I could do to tell it to run the local impl vs using the remote one (I can do the impl-choosing logic myself, just not sure how to spell what I want).

Also, why did these tests not fail if I am calling this incorrectly?

Reply:
@azaslavsky Because the import tests are not marked with either `@region_silo_test(stable=True)` or `@region_silo_test(stable=True)` to establish the expected silo. Without these decorators, the test suite runs in monolith silo mode. Hence the tests won't fail.

I've added this snippet into `import_by_model()`, which caused some tests in `tests/sentry/backup/test_imports.py` to fail:

```python
        if in_test_environment():
            silo_limit = getattr(model._meta, "silo_limit", None)
            if not silo_limit:
                raise Exception("Expected silo_limit")
            if SiloMode.CONTROL not in silo_limit.modes:
                raise Exception("Accessed non-control model")
```

Reply:
I've reworked this PR to work over RPC properly, with the `stable=True` annotations placed as needed.